File: go800mon/a800mon/monitor/topbar.go
package monitor

import (
	"context"
	"fmt"

	. "go800mon/a800mon"
)

const (
	topbarTitle      = "Atari800 Monitor"
	topbarCopyright  = "(c) 2026 Marcin Nowak"
	topbarRightWidth = 43
)

type TopBar struct {
	BaseWindowComponent
	lastSnapshot string
}

type topbarSegment struct {
	text  string
	color Color
}

func NewTopBar(window *Window) *TopBar {
	return &TopBar{BaseWindowComponent: NewBaseWindowComponent(window)}
}

func (t *TopBar) Update(_ctx context.Context) (bool, error) {
	st := State()
	snap := fmt.Sprintf("%s|%t|%d|%d|%d|%t", st.LastRPCError, st.Crashed, st.EmuMS, st.ResetMS, st.MonitorFrameTimeMS, st.UIFrozen)
	if t.lastSnapshot == snap {
		return false, nil
	}
	t.lastSnapshot = snap
	return true, nil
}

func (t *TopBar) Render(_force bool) {
	st := State()
	w := t.Window()
	w.Cursor(0, 0)
	if st.LastRPCError != "" {
		w.Print(topbarTitle+" ", ColorTopbar.Attr(), false)
		w.Print(" "+st.LastRPCError+" ", ColorError.Attr(), false)
		w.FillToEOL(' ', ColorError.Attr())
	} else {
		w.Print(topbarTitle+"     "+topbarCopyright, ColorTopbar.Attr(), false)
		if st.UIFrozen {
			w.Print("   ", ColorTopbar.Attr(), false)
			w.Print(" FREEZE ", ColorError.Attr(), false)
		}
		w.FillToEOL(' ', ColorTopbar.Attr())
	}
	start := w.Width() - topbarRightWidth
	if start < 0 {
		start = 0
	}
	w.Cursor(start, 0)
	segments := []topbarSegment{
		{text: crashLabel(st.Crashed), color: crashColor(st.Crashed)},
		{text: " UP ", color: ColorText},
		{text: fmt.Sprintf(" %s ", FormatHMS(st.EmuMS)), color: ColorTopbar},
		{text: " RS ", color: ColorText},
		{text: fmt.Sprintf(" %s ", FormatHMS(st.ResetMS)), color: ColorTopbar},
		{text: fmt.Sprintf(" %3d ms ", st.MonitorFrameTimeMS), color: ColorText},
	}
	for _, segment := range segments {
		w.Print(segment.text, segment.color.Attr(), false)
	}
}

func crashLabel(crashed bool) string {
	if crashed {
		return " CRASH "
	}
	return "       "
}

func crashColor(crashed bool) Color {
	if crashed {
		return ColorError
	}
	return ColorTopbar
}

File: py800mon/monitor/topbar.py
from ..app import Component, VisualComponent
from ..ui import Color
from .appstate import state

CRASH_LABEL = {
    True: (" CRASH ", Color.ERROR),
    False: ("       ", Color.TOPBAR),
}
TITLE = "Atari800 Monitor"
COPYTIGHT = "(c) 2026 Marcin Nowak"
TOPRIGHT_LEN = 43


class TopBar(VisualComponent):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._last_snapshot = None

    async def update(self):
        snapshot = (
            state.last_rpc_error,
            state.crashed,
            state.emu_ms,
            state.reset_ms,
            state.monitor_frame_time_ms,
            state.ui_frozen,
        )
        if self._last_snapshot == snapshot:
            return False
        self._last_snapshot = snapshot
        return True

    def render(self, force_redraw=False):
        self.window.cursor = 0, 0
        if state.last_rpc_error:
            self.window.print(f"{TITLE} ", Color.TOPBAR.attr())
            self.window.print(f" {state.last_rpc_error} ", Color.ERROR.attr())
            self.window.fill_to_eol(attr=Color.ERROR.attr())
        else:
            self.window.print(f"{TITLE}     {COPYTIGHT}", Color.TOPBAR.attr())
            if state.ui_frozen:
                self.window.print("   ", Color.TOPBAR.attr())
                self.window.print(" FREEZE ", Color.ERROR.attr())
            self.window.fill_to_eol(attr=Color.TOPBAR.attr())

        segments = (
            CRASH_LABEL[state.crashed],
            (" UP ", Color.TEXT),
            (f" {_format_hms(state.emu_ms)} ", Color.TOPBAR),
            (" RS ", Color.TEXT),
            (f" {_format_hms(state.reset_ms)} ", Color.TOPBAR),
            (f" {state.monitor_frame_time_ms:3d} ms ", Color.TEXT),
        )
        self.window.cursor = (self.window._iw - TOPRIGHT_LEN, 0)
        for text, color in segments:
            self.window.print(text, attr=color.attr())


def _format_hms(ms):
    total = max(0, int(ms // 1000))
    h = total // 3600
    m = (total % 3600) // 60
    s = total % 60
    return f"{h:02d}:{m:02d}:{s:02d}"
